File: backend/Questionnaire/Questionnaire.Contract/Models/Questions/Question.cs
using System.Text.Json.Serialization;

namespace Questionnaire.Contract.Models;

[JsonPolymorphic(TypeDiscriminatorPropertyName = "questionType")]
[JsonDerivedType(typeof(MessageQuestion), "Message")]
[JsonDerivedType(typeof(TextQuestion), "Text")]
[JsonDerivedType(typeof(NumberQuestion), "Number")]
[JsonDerivedType(typeof(EmailQuestion), "Email")]
public abstract class Question
{
    public required QuestionId Id { get; init; } = QuestionId.New();
    public required string Title { get; set; }
    public string? Description { get; set; }
    public List<Question> SubQuestions { get; set; } = new();

    [JsonIgnore]
    public abstract string QuestionType { get; }
}

[JsonPolymorphic(TypeDiscriminatorPropertyName = "questionType")]
[JsonDerivedType(typeof(MessageQuestionDelta), "Message")]
[JsonDerivedType(typeof(TextQuestionDelta), "Text")]
[JsonDerivedType(typeof(NumberQuestionDelta), "Number")]
[JsonDerivedType(typeof(EmailQuestionDelta), "Email")]
public class QuestionDelta
{
    public required QuestionId Id { get; init; }
    public Patchable<string> Title { get; set; } = Patchable<string>.NotGiven();
    public Patchable<string?> Description { get; set; } = Patchable<string?>.NotGiven();
    public PatchableArray<QuestionDelta, QuestionId>? SubQuestions { get; set; }

    public void Apply(Question question)
    {
        if (question.Id != Id)
            throw new InvalidOperationException("Delta Id does not match Question Id");

        Title.Apply(question, (q, v) => { if (v != null) q.Title = v; });
        Description.Apply(question, (q, v) => q.Description = v);

        if (SubQuestions != null)
        {
            ApplyQuestionDeltaPatchableArray(question.SubQuestions, SubQuestions);
        }
    }

    private void ApplyQuestionDeltaPatchableArray(List<Question> list, PatchableArray<QuestionDelta, QuestionId> patch)
    {
        switch (patch.Operation)
        {
            case PatchableArrayOperation.Replace:
            case PatchableArrayOperation.Add:
            case PatchableArrayOperation.AddRange:
                throw new NotSupportedException("Add/Replace operations not supported for QuestionDelta patches. Use full Question objects.");

            case PatchableArrayOperation.Remove:
                if (patch.Index.HasValue && patch.Index.Value >= 0 && patch.Index.Value < list.Count)
                    list.RemoveAt(patch.Index.Value);
                break;

            case PatchableArrayOperation.RemoveById:
                if (patch.ItemId.HasValue)
                {
                    var question = list.FirstOrDefault(q => q.Id == patch.ItemId.Value);
                    if (question != null)
                        list.Remove(question);
                }
                break;

            case PatchableArrayOperation.RemoveRange:
                if (patch.Index.HasValue && patch.Count.HasValue &&
                    patch.Index.Value >= 0 && patch.Index.Value < list.Count)
                {
                    var count = Math.Min(patch.Count.Value, list.Count - patch.Index.Value);
                    list.RemoveRange(patch.Index.Value, count);
                }
                break;

            case PatchableArrayOperation.Move:
                if (patch.Index.HasValue && patch.ToIndex.HasValue &&
                    patch.Index.Value >= 0 && patch.Index.Value < list.Count &&
                    patch.ToIndex.Value >= 0 && patch.ToIndex.Value < list.Count)
                {
                    var item = list[patch.Index.Value];
                    list.RemoveAt(patch.Index.Value);
                    list.Insert(patch.ToIndex.Value, item);
                }
                break;

            case PatchableArrayOperation.MoveById:
                if (patch.ItemId.HasValue && patch.ToIndex.HasValue)
                {
                    var question = list.FirstOrDefault(q => q.Id == patch.ItemId.Value);
                    if (question != null)
                    {
                        list.Remove(question);
                        var toIndex = Math.Min(patch.ToIndex.Value, list.Count);
                        list.Insert(toIndex, question);
                    }
                }
                break;

            case PatchableArrayOperation.Clear:
                list.Clear();
                break;
        }

        // After structural operations, apply any deltas to existing questions
        if (patch.Item != null)
        {
            var existing = list.FirstOrDefault(q => q.Id == patch.Item.Id);
            if (existing != null)
                patch.Item.Apply(existing);
        }

        if (patch.Items != null)
        {
            foreach (var delta in patch.Items)
            {
                var existing = list.FirstOrDefault(q => q.Id == delta.Id);
                if (existing != null)
                    delta.Apply(existing);
            }
        }
    }
}

File: client-ts/model/question.ts
import { Patchable, PatchableNullable, PatchableArray } from './delta';

export type QuestionType = 'Message' | 'Text' | 'Number' | 'Email';

export interface QuestionBase {
  id: string;
  title: string;
  description?: string | null;
  subQuestions: Question[];
  questionType: QuestionType;
}

export interface MessageQuestion extends QuestionBase {
  questionType: 'Message';
}

export interface TextQuestion extends QuestionBase {
  questionType: 'Text';
}

export interface NumberQuestion extends QuestionBase {
  questionType: 'Number';
}

export interface EmailQuestion extends QuestionBase {
  questionType: 'Email';
}

export type Question = MessageQuestion | TextQuestion | NumberQuestion | EmailQuestion;

export class QuestionDelta {
  id: string;
  title: Patchable<string>;
  description: PatchableNullable<string>;
  subQuestions?: PatchableArray<QuestionDelta>;

  constructor(id: string) {
    this.id = id;
    this.title = Patchable.notGiven<string>();
    this.description = PatchableNullable.notGiven<string>();
  }

  apply(question: Question): void {
    if (question.id !== this.id) {
      throw new Error('Delta Id does not match Question Id');
    }

    this.title.apply(question, (q, v) => {
      q.title = v;
    });
    this.description.apply(question, (q, v) => {
      q.description = v;
    });

    if (this.subQuestions) {
      applyQuestionDeltaPatchableArray(question.subQuestions, this.subQuestions);
    }
  }
}

export interface MessageQuestionDelta extends QuestionDelta {
  questionType: 'Message';
}

export interface TextQuestionDelta extends QuestionDelta {
  questionType: 'Text';
}

export interface NumberQuestionDelta extends QuestionDelta {
  questionType: 'Number';
}

export interface EmailQuestionDelta extends QuestionDelta {
  questionType: 'Email';
}

export function applyQuestionDeltaPatchableArray(
  list: Question[],
  patch: PatchableArray<QuestionDelta>
): void {
  switch (patch.operation) {
    case 'Replace':
    case 'Add':
    case 'AddRange':
      throw new Error('Add/Replace operations not supported for QuestionDelta patches. Use full Question objects.');

    case 'Remove':
      if (patch.index != null && patch.index >= 0 && patch.index < list.length) {
        list.splice(patch.index, 1);
      }
      break;

    case 'RemoveById':
      if (patch.itemId != null) {
        const index = list.findIndex((q) => q.id === patch.itemId);
        if (index >= 0) {
          list.splice(index, 1);
        }
      }
      break;

    case 'RemoveRange':
      if (
        patch.index != null &&
        patch.count != null &&
        patch.index >= 0 &&
        patch.index < list.length
      ) {
        const count = Math.min(patch.count, list.length - patch.index);
        list.splice(patch.index, count);
      }
      break;

    case 'Move':
      if (
        patch.index != null &&
        patch.toIndex != null &&
        patch.index >= 0 &&
        patch.index < list.length &&
        patch.toIndex >= 0 &&
        patch.toIndex < list.length
      ) {
        const [item] = list.splice(patch.index, 1);
        list.splice(patch.toIndex, 0, item);
      }
      break;

    case 'MoveById':
      if (patch.itemId != null && patch.toIndex != null) {
        const index = list.findIndex((q) => q.id === patch.itemId);
        if (index >= 0) {
          const [item] = list.splice(index, 1);
          const toIndex = Math.min(patch.toIndex, list.length);
          list.splice(toIndex, 0, item);
        }
      }
      break;

    case 'Clear':
      list.splice(0, list.length);
      break;
  }

  // After structural operations, apply any deltas to existing questions
  if (patch.item != null) {
    const existing = list.find((q) => q.id === patch.item!.id);
    if (existing) {
      patch.item.apply(existing);
    }
  }

  if (patch.items != null) {
    for (const delta of patch.items) {
      const existing = list.find((q) => q.id === delta.id);
      if (existing) {
        delta.apply(existing);
      }
    }
  }
}
